### ISSUE TITLE
Colorbar fill not aligned with containing box

### ISSUE THREAD

**Issue by brunobeltran** [matplotlib/matplotlib#17103]
### Bug report

**Bug summary**

Definitely a low-priority bug, but at some figure sizes, there seems to be an issue with the colorbar not being drawn at the correct position in some backends (seems like PDF and SVG are affected, but not Agg).

The colorbar "fill" (the actual colored bit) will not line up with the box that contains it (the black marker edge that makes the rectangle that is filled by the color bar, see below). Seems like some sort of rounding issue.

I have a more "dramatic" example, but still need to distill it to a minimum working example, as it is a plot of proprietary data.

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import matplotlib.cm

fig, ax = plt.subplots(figsize=(2, 1.5))

plt.plot()

sm = matplotlib.cm.ScalarMappable()
sm.set_array([])
plt.colorbar(sm)

plt.savefig('/home/bbeltr1/Downloads/test.svg')
plt.savefig('/home/bbeltr1/Downloads/test.pdf')
plt.savefig('/home/bbeltr1/Downloads/test.png', dpi=800)

```

**Actual outcome**

SVG:
![test svg](https://user-images.githubusercontent.com/1475390/79055002-608d3d00-7bfe-11ea-9487-fa3c80535bfd.png)

PDF:
![test pdf](https://user-images.githubusercontent.com/1475390/79055003-62570080-7bfe-11ea-8de0-c6afad7dc7ab.png)

PNG:
![test](https://user-images.githubusercontent.com/1475390/79055004-6420c400-7bfe-11ea-953f-6b3071b74244.png)

**Expected outcome**

The PNG output has the expected outcome.

**Matplotlib version**

  * Operating system: Debian Jessie
  * Matplotlib version: master
  * Matplotlib backend (`print(matplotlib.get_backend())`):
  * Python version: 3.7
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by ImportanceOfBeingErnest** [comment]
Might this be a problem with the viewer? 

For me, pdf looks like this in Foxit Viewer:

![image](https://user-images.githubusercontent.com/23121882/79055313-6ad62800-7c4c-11ea-9fd5-8c1f77db6c0a.png)
 and svg like this in Inkscape:

![image](https://user-images.githubusercontent.com/23121882/79055319-7cb7cb00-7c4c-11ea-99c0-093a3c472c22.png)

and like this in Firefox:

![image](https://user-images.githubusercontent.com/23121882/79055342-95c07c00-7c4c-11ea-8f01-7d8394e629d1.png)

**Comment by jklymak** [comment]
I can't reproduce in Preview nor Adobe Acrobat (on Mac).

**Comment by brunobeltran** [comment]
Interesting, I can reproduce it in Chrome but by default "qpdfview" doesn't show the problem until I zoom in. Chrome looks like the below (smaller gap than in Inkscape, but still a gap) in both PDF and SVG


![2020-04-12-194859_613x259_scrot](https://user-images.githubusercontent.com/1475390/79088203-cbba3a80-7cf6-11ea-9beb-5ba24d6e3a65.png)

Whereas "qpdfview" looks like this:
![2020-04-12-195128_887x668_scrot](https://user-images.githubusercontent.com/1475390/79088289-0623d780-7cf7-11ea-8cfa-90b14cf861fb.png)

Until I zoom REALLY far in, then it looks like 
![2020-04-12-195158_818x478_scrot](https://user-images.githubusercontent.com/1475390/79088319-1936a780-7cf7-11ea-9351-586393794fde.png)

**Comment by brunobeltran** [comment]
Also Foxit on my system, still shows the alignment off. But you can only see it if you zoom in a bit.

![2020-04-12-234607_1475x660_scrot](https://user-images.githubusercontent.com/1475390/79098952-ef41ad00-7d17-11ea-9ba2-35ffdc74ffa1.png)

**Comment by brunobeltran** [comment]
> I can't reproduce in Preview nor Adobe Acrobat (on Mac).

@jklymak My roommate on a mac can reproduce this by zooming into the bottom of the bar in Preview. The gap renders only after a certain zoom level.

**Comment by timhoffm** [comment]
Maybe a similar rendering artifact like #5694.

**Comment by brunobeltran** [comment]
> Maybe a similar rendering artifact like #5694.

I suspect not actually, because I can make a better example that also "overflows" the top of the bar instead of just "underflowing" the top. Sorry that it took awhile to get this:

**Better Example**

```
import matplotlib.pyplot as plt
from matplotlib.transforms import Bbox

# [ax.get_position() for ax in proprietary_fig.axes]
ax_bbox = Bbox([[0.1974965228858249, 0.203249633462262], [0.8130847727044617, 0.8860326455421427]])
cax_bbox = Bbox([[0.8515590383181265, 0.203249633462262], [0.8753374421595627, 0.8860326455421427]])

# figsize=proprietary_fig.get_size_inches()
fig = plt.figure(figsize=[2.27, 1.58108728])

ax = fig.add_axes([ax_bbox.x0, ax_bbox.y0, ax_bbox.width, ax_bbox.height])
cax = fig.add_axes([cax_bbox.x0, cax_bbox.y0, cax_bbox.width, cax_bbox.height])

cb = plt.colorbar(sm, cax=cax)

plt.savefig('/home/bbeltr1/Downloads/test-no-plot-no-labels.svg')
```

**Output**

![2020-04-13-022012_114x474_scrot](https://user-images.githubusercontent.com/1475390/79109583-7cdbc780-7d2d-11ea-8411-5d78d6be0bba.png)

**Comment by brunobeltran** [comment]
For the above colorbar, opening the SVG shows that the "box" ends up being called `<svg:g id="patch_8">` and is drawn as 

```
M 139.17881,90.700695 V 90.397074 13.277469 12.973848 h 3.88634 v 0.303621 77.119605 0.303621 z
```

which is functionally equivalent to just

```
M 139.17881,90.700695 V 12.973848 h 3.88634 v 77.726847 z
```


whereas the image is a...well....`svg:image` with `x=139`, `y=-12`, `width=4`, `height=78` (and `transform=matrix(1, 0, 0, -1, 0, 78)`).

This suggests the issue is actually the PNG position is being rounded to the nearest pixel, whereas the SVG path data is being kept as floats. 

By zooming in, it is clear that the bar is slightly to the left (139 instead of 139.17...), is slightly to wide (4 instead of 3.886...), etc.

**Comment by ImportanceOfBeingErnest** [comment]
When I run the new example code, I can still not reproduce the issue. I do get 

    transform="scale(1 -1)translate(0 -77.76)" height="77.76" width="4.32" x="138.96" y="-12.96"

in the svg file.

Here are the two files I obtain (zipped, for github to me upload them):
[test.zip](https://github.com/matplotlib/matplotlib/files/4469169/test.zip)

**Comment by brunobeltran** [comment]
Okay, thanks everyone for your patience here, after games of reinstalling-everything-in-a-fresh-venv , I think I've determined that this is a Jupyter-specific bug. 

The following produces a "correct" plot, as reported by @ImportanceOfBeingErnest:

```
$ python -m venv venv
$ source venv/bin/activate
$ pip install matplotlib
$ cat make_plot.py
import matplotlib.cm
import matplotlib.pyplot as plt
from matplotlib.transforms import Bbox

# [ax.get_position() for ax in proprietary_fig.axes]
ax_bbox = Bbox([[0.1974965228858249, 0.203249633462262], [0.8130847727044617, 0.8860326455421427]])
cax_bbox = Bbox([[0.8515590383181265, 0.203249633462262], [0.8753374421595627, 0.8860326455421427]])

# figsize=proprietary_fig.get_size_inches()
fig = plt.figure(figsize=[2.27,1.58108728])

ax = fig.add_axes([ax_bbox.x0, ax_bbox.y0, ax_bbox.width, ax_bbox.height])
cax = fig.add_axes([cax_bbox.x0, cax_bbox.y0, cax_bbox.width, cax_bbox.height])

sm = matplotlib.cm.ScalarMappable()
sm.set_array([])

cb = plt.colorbar(sm, cax=cax)

plt.savefig('./test-plot.svg')
$ python make_plot.py
```

However, opening a new jupyter notebook and executing 
```
%run make_plot.py
```
produces a PDF with the rounding issue.

**Comment by ImportanceOfBeingErnest** [comment]
Ok, this does not mean that it's a bug with jupyter necessarily. After all, it's still matplotlib producing the output. But most probably there are some settings being made (e.g. if using IPykernel's inline backend) that change the numbers. It might then be possible to construct an example which shows the same problem outside of jupyter as well.

**Comment by QuLogic** [comment]
Looks like #6827?

**Comment by brunobeltran** [comment]
Jupyter does set the DPI itself actually....maybe I'll dig into this, seems like it is probably the same issue.

**Comment by cover-me** [comment]
> For the above colorbar, opening the SVG shows that the "box" ends up being called `<svg:g id="patch_8">` and is drawn as
> 
> ```
> M 139.17881,90.700695 V 90.397074 13.277469 12.973848 h 3.88634 v 0.303621 77.119605 0.303621 z
> ```
> 
> which is functionally equivalent to just
> 
> ```
> M 139.17881,90.700695 V 12.973848 h 3.88634 v 77.726847 z
> ```
> 
> whereas the image is a...well....`svg:image` with `x=139`, `y=-12`, `width=4`, `height=78` (and `transform=matrix(1, 0, 0, -1, 0, 78)`).
> 
> This suggests the issue is actually the PNG position is being rounded to the nearest pixel, whereas the SVG path data is being kept as floats.
> 
> By zooming in, it is clear that the bar is slightly to the left (139 instead of 139.17...), is slightly to wide (4 instead of 3.886...), etc.

I can reproduce the problem with Chrome + Jupyter notebook.  Matplotlib version 3.2.0.

The SVG result:

![Screenshot 2020-04-15 at 18 34 08](https://user-images.githubusercontent.com/22870592/79395532-d4557f80-7f47-11ea-835e-eb64210ee11c.png)

However, the white space goes away after I set a larger figure size (2*1.5->3*2.25)

![Screenshot 2020-04-15 at 18 38 07](https://user-images.githubusercontent.com/22870592/79395779-665d8800-7f48-11ea-8bbb-7f94e1ed6570.png)

Maybe this is an issue due to rounding.

**Comment by cover-me** [comment]
Image positions are rounded to 72/dpi in SVG files. The default dpi in Matplotlib is 100 or so. For Jupyter notebook, dpi=72, so the position is rounded to integers. Increase the DPI would increase the precision of positions. `plt.savefig('test.svg')` -> `plt.savefig('test.svg',dpi=100)`

**Comment by jklymak** [comment]
Interesting - do the images need to be snapped like that?

**Comment by QuLogic** [comment]
I'm fairly certain this is a duplicate of #6827 then.

**Comment by drcege** [comment]
Thank you all! Manually setting `dpi` in the `savefig` function solved the issue.
